syntax = "proto3";

package proto;

option java_package = "edu.bu.tweet.proto";
option java_outer_classname = "TweetStatusProtos";

message TweetStatusP {

    string createdAt = 1;
    int64 id = 2;
    string text = 3;
    string source = 4;
    bool isTruncated = 5;
    int64 inReplyToStatusId = 6;
    int64 inReplyToUserId = 7;
    string inReplyToScreenName = 8;

    //User:
    message UserP {

        int64 id = 1;
        string name = 2;
        string screenName = 3;
        string location = 4;
        string url = 5;
        string description = 6;
        bool isProtected = 7;
        bool isVerified = 8;
        int32 followersCount = 9;
        int32 friendsCount = 10;
        int32 listedCount = 11;
        int32 favouritesCount = 12;
        int32 statusesCount = 13;
        string createdAt = 14;
        string profileBannerImageUrl = 15;
        string profileImageUrlHttps = 16;
        bool isDefaultProfile = 17;
        repeated string withheldInCountries = 18;
        string withheldScope = 19;
        repeated URLEntityP descriptionURLEntities = 20;
        bool isGeoEnabled = 21;
        string lang = 22;
        bool isContributorsEnabled = 23;
        string profileBackgroundColor = 24;
        string profileBackgroundImageUrl = 25;
        string profileBackgroundImageUrlHttps = 26;
        bool profileBackgroundTiled = 27;
        string profileImageUrl = 28;
        string profileLinkColor = 29;
        string profileSidebarBorderColor = 30;
        string profileSidebarFillColor = 31;
        string profileTextColor = 32;
        bool profileUseBackgroundImage = 33;
        bool isDefaultProfileImage = 34;
        int32 utcOffset = 35;
        string timeZone = 36;
        bool translator = 37;
        bool isFollowRequestSent = 38;
        bool showAllInlineMedia = 39;
    }

    UserP userP = 27;

    //Coordinates
    message CoordinatesP {
        string type = 1;
        repeated double coordinates = 2;
    }
    CoordinatesP coordinates = 28;

    //Place
    message PlaceP {
        string name = 1;
        string countryCode = 2;
        string id = 3;
        string country = 4;
        string placeType = 5;
        string url = 6;
        string fullName = 7;

        //BoundingBoxCoordinate
        message BoundingBoxCoordinateP {
            string type = 1;

            // Representation of a three dimensional array for coordinates
            message RowsOneP {
                message RowsTwoP {
                    repeated double coordinate = 1;
                }
                repeated RowsTwoP rowsTwoP = 1;
            }
            repeated RowsOneP threeDimArray = 2;

        }

        BoundingBoxCoordinateP boundingBoxCoordinateP = 8;
    }
    PlaceP place = 29;
    int64 quotedStatusId = 30;
    bool isQuoteStatus = 31;
    TweetStatusP quotedStatus = 32;
    TweetStatusP retweetedStatus = 33;
    int32 quoteCount=34;
    int32 replyCount=35;
    int32 retweetCount=36;
    int32 favoriteCount=37;

    //Entities
    message EntitiesP{
        repeated HashtagEntityP hashtagEntityP=1;
        repeated MediaEntityP mediaEntityP=2;
        repeated URLEntityP urlEntityP = 3;
        repeated UserMentionEntityP userMentionEntityP = 4;
        repeated SymbolEntityP symbolEntityP = 5;
        repeated PollEntityP pollEntityP = 6;

    }
    EntitiesP entitiesP=38;

    message ExtendedEntitiesP{
        repeated MediaEntityP media=1;
    }

    ExtendedEntitiesP extendedEntities=39;

    bool isFavorited=40;
    bool isRetweeted=41;
    bool isPossiblySensitive=42;
    string filterLevel=43;
    string lang=44;
    repeated MatchingRulesEntityP matchingRules=45;
    int64 currentUserRetweetedId=46;
    repeated MapFieldEntry scopes=47;
    bool withheldCopyright=48;
    repeated string withheldInCountries=49;
    string  withheldScope=50;


}

//URLEntityP
message URLEntityP {
    repeated int32 indices = 1;
    string displayURL = 2;
    string expandedURL = 3;
    string url = 4;
}

//HashtagEntity
message HashtagEntityP{
    repeated int32 indices=1;
    string text=2;
}
//MediaEntity
message MediaEntityP{
    string displayURL=1;
    string expandedURL=2;
    int64 id=3;
    repeated int32 indices=4;
    string mediaURL=5;
    string mediaURLHttps=6;
    MediaSizesEntityP sizes=7;
    string type=8;
    int64 sourceStatusId=9;
    string sourceStatusIdStr=10;
    string url=11;
    VideoEntityP videoInfo=12;
    AdditionalMediaInfoEntityP additionalMediaInfo=13;
}

//UserMentionEntity
message UserMentionEntityP{
    int64 id=1;
    string idStr=2;
    repeated int32 indices=3;
    string name=4;
    string screenName=5;
}

//SymbolEntity
message SymbolEntityP{
    repeated int32 indices=1;
    string text=2;
}

//PollEntity
message PollEntityP{
    repeated OptionEntityP options=1;
    string endDatetime=2;
    string durationMinutes=3;
}

//MediaSizesEntity
message MediaSizesEntityP{
    SizeEntityP thumb=1;
    SizeEntityP large=2;
    SizeEntityP medium=3;
    SizeEntityP small=4;
}

//SizeEntity
message SizeEntityP{
    int32 width=1;
    int32 height=2;
    string resize=3;
}

//VariantEntity
message VariantEntityP{
    int64 bitrate=1;
    string contentType=2;
    string url=3;
}
//VideoEntity
message VideoEntityP{
    repeated int32 aspectRatio=1;
    int32 durationMillis=2;
    repeated VariantEntityP variants=3;
}

//AdditionalMediaInfoEntity
message AdditionalMediaInfoEntityP{
    string title=1;
    string description=2;
    bool embeddable=3;
    bool monetizable=4;
}

//OptionEntity
message OptionEntityP{
    int32 position=1;
    string text=2;
}

//MatchingRulesEntity
message MatchingRulesEntityP{
    string tag=1;
    int64 id=2;
    string idStr=3;
}

//Map field
message MapFieldEntry {
    string key = 1;
    bool value = 2;
}









